distribution/openrct2.d.ts

```diff
@@ -3150,6 +3151,118 @@ declare global {
         postMessage(message: ParkMessageDesc): void;
     }
 
+    interface Research {
+        /**
+         * The list of rides and scenery sets that have already been researched.
+         */
+        inventedItems: ResearchItem[];
+
+        /**
+         * The order of rides and scenery sets to be researched.
+         */
+        uninventedItems: ResearchItem[];
+
+        /**
+         * The amount of funding currently spent on research.
+         */
+        funding: ResearchFundingLevel;
+
+        /**
+         * Flags representing which research categories are enabled.
+         */
+        priorities: number;
+
+        /**
+         * The current stage for the ride or scenery set being researched.
+         */
+        stage: ResearchFundingStage;
+
+        /**
+         * The progress for the current stage between 0 and 65535.
+         * This will increment more quickly the higher the research funding.
+         */
+        progress: number;
+
+        /**
+         * The expected month the current item being researched will complete.
+         * Value is between 0 and 7, 0 being March and 7 being October.
+         * Value is null if there is not yet an expected month.
+         */
+        readonly expectedMonth: number | null;
+
+        /**
+         * The expected day of the month the current item being researched will complete.
+         * Value is between 0 and 30, add 1 to it for the human readable date.
+         * Value is null if there is not yet an expected month.
+         */
+        readonly expectedDay: number | null;
+
+        /**
+         * Gets whether a particular object has been researched and is available to construct.
+         * @param type The type of object, e.g. ride, scenery group, or small scenery.
+         * @param index The object index.
+         */
+        isObjectResearched(type: ObjectType, index: number): boolean;
+    }
+
+    type ResearchItem = RideResearchItem | SceneryResearchItem;
+
+    interface RideResearchItem {
+        type: "ride";
+
+        /**
+         * The research category this item belongs in.
+         * E.g. gentle rides, thrill rides, shops etc.
+         */
+        category?: ResearchCategory;
+
+        /**
+         * The ride type. Each vehicle can have a seperate invention for each ride type.
+         */
+        rideType: number;
+
+        /**
+         * The ride (vehicle) object index.
+         */
+        object: number;
+    }
+
+    interface SceneryResearchItem {
+        category?: ResearchCategory.SceneryGroup;
+        type: "scenery";
+
+        /**
+         * The scenery set object index.
+         */
+        object: number;
+    }
+
+    type ResearchItemType = "scenery" | "ride";
+
+    enum ResearchCategory {
```

Comment:
The problem with enums is that e.g. `ResearchCategory.Transport` will not throw an error during transpilation, but during runtime.
I would prefer types as is done with e.g. ResearchFundingStage.

Reply:
I wanted to make an exception for this as the category will be ignored by the game, so plugins won't need to provide it. It also optimises the large list of research items as it does not need to allocate a string for each item.

Reply:
Maybe, I don't quite know how to do this if you can. But the original reason @StephanSpengler mentioned against this doesn't apply here.

Reply:
I think a string union would be better here as well. It would make it consistent with how tile elements are unioned.

If it helps, I did find the Duktape docs about how it handles strings: https://wiki.duktape.org/performance#duktape-performance-characteristics

Reply:
Okay the other reason for using an enum here is that you need them to use the priorities property which is a bit mask using these values.

Reply:
In this API there are tons of other numeric values with semantic meaning that do not have an enum. I really think we should not use them, it creates too many problems.

Reply:
The bit mask wouldn't really work in the current implementation either as they're not numbered yet; `ResearchCategory.Thrill` is currently number 3 and would thus equal `ResearchCategory.Gentle | ResearchCategory.Rollercoaster` (numbered 1 and 2).

Reply:
What you would do is `value = (1 << enumValue1) | (1 << enumValue2) | ... ;

Reply:
That sounds very unnecessarily convoluted... 😅 IMHO the API should take an approach that's easier to use and more beginner friendly than that. And make it a bit consistent with what we already have now.